docs(css): FF127 - `calc()` support for color channels (#34046)
mdn/content: files/en-us/web/css/calc/index.md

{{CSSRef}}

The **`calc()`** [CSS](/en-US/docs/Web/CSS) [function](/en-US/docs/Web/CSS/CSS_Functions) lets you perform calculations when specifying CSS property values. It can be used with {{cssxref("&lt;length&gt;")}}, {{cssxref("&lt;frequency&gt;")}}, {{cssxref("&lt;angle&gt;")}}, {{cssxref("&lt;time&gt;")}}, {{cssxref("&lt;percentage&gt;")}}, {{cssxref("&lt;number&gt;")}}, {{cssxref("&lt;integer&gt;")}}, and {{cssxref("color_value", "&lt;color-function&gt;")}} values.

{{EmbedInteractiveExample("pages/css/function-calc.html")}}

## Syntax

```css
/* calc(expression) */
calc(100% - 80px)

/* Expression with a CSS function */
calc(100px * sin(pi / 2))

/* Expression containing a variable */
calc(var(--hue) + 180)

/* Expression with color channels in relative colors */

> rgb(from blue calc(r * .3) calc(g * .59) calc(b * .11))

Reviewer: blue is rgb(0 0 255 / 1 ) 
The calc here is  basically rgb((0*.3) (0 * .59) and (255 * .11)) or 
rgb(0 0 28).
But since your using an old color function with a limited gamut, the output here is `color(srgb 0 0 0.11)`. 

Suggest using a relative color using a function that doesn't return a `color()` function instead of the color function used. 
Also suggest using calc on one to two calcs, as most people will calc a hue or an alpha, but not change all the color channels. Even more useful doing a relative alpha or hue on a color custom property, but that might be beyond the scope of this page.

> rgb(from blue calc(r * .3) calc(g * .59) calc(b * .11))

Reviewer 2: What about using lch()?

```suggestion
lch(from aquamarine l c calc(h + 180))
```

Could probably duplicate it below, also

## Syntax

```css
/* calc(expression) */

> calc(100% - 80px)

Reviewer 2: Should we add this:

```suggestion
calc(100% - 80px)

/* Expression with a CSS function */
calc(100px * sin(pi / 2))

/* Expression containing a variable */
calc(var(--hue) + 180)
```

Author: This is great, thanks! Adding

> /* Expression with color channel */
> calc(l * 0.8)
> calc(h + 180)
> calc(b / 2)
> calc(r * .3 + g * .59 + b * .11)

Reviewer 2: What do you think about this?

```suggestion
/* Expression with color channels */
calc(alpha * 0.8)
calc(r / 2) calc(g + 25) calc(b + 175)
```

For the last one, I think it needs to be in this format:

```css
  background-color: rgb(from blue calc(r * .3) calc(g * .59) calc(b * .11));
```

Reviewer 2: And we're talking about use in relative colors, so this is more explicit, but it's verbose

```
/* Expression with color channels in relative colors */
```

Author: - The "Syntax" guidelines for CSS function pages were updated: https://developer.mozilla.org/en-US/docs/MDN/Writing_guidelines/Page_structures/Page_types/CSS_function_page_template#syntax, and so now we don't mention the property in which the function is used or the ending semi-colon. I realize there are loads of other CSS function pages where we still need to implement this new guideline.

- Agree with your suggestion about explicitly stating use in relative colors

- For lengths, you can't use `0` to mean `0px` (or another length unit); instead, you must use the version with the unit: `margin-top: calc(0px + 20px);` is valid, while `margin-top: calc(0 + 20px);` is invalid.
- The `calc()` function cannot directly substitute the numeric value for percentage types; for instance `calc(100 / 4)%` is invalid, while `calc(100% / 4)` is valid.
- Current implementations require that for the `*` and `/` operators, one of the operands has to be unitless. For `/`, the right operand must be unitless. For example `font-size: calc(1.25rem / 1.25)` is valid but `font-size: calc(1.25rem / 125%)` is invalid.

> - When `calc()` is used where an {{cssxref("&lt;integer&gt;")}} is expected, the value will be rounded to the nearest integer. For example, `z-index: calc(3 / 2);` will result in a `z-index` value of `2`.

Reviewer 2: I guess it's rounded up, is it?

Reviewer 2: Well, I guess rounded to the nearest integer is fine, but if halfway, it's rounded up:

* `z-index: calc(1.4);` = 1
* `z-index: calc(1.5);` = 2
* `z-index: calc(1.6);` = 2

Author: Good point. I also found this reference in the spec: [calc type checking](https://drafts.csswg.org/css-values/#calc-type-checking) leading to [round to the nearest integer](https://drafts.csswg.org/css-values/#css-round-to-the-nearest-integer).
I've updated the text to (removed z-index from the example as it was distracting):

> - When `calc()` is used where an {{cssxref("&lt;integer&gt;")}} is expected, the value will be rounded to the nearest integer. So, `calc(1.4)` will result in a value of `1`. If the fractional part of the value is exactly `0.5`, the value will be rounded up. For example, `calc(1.5)` will result in a value of `2`, while `calc(-1.5)` will round to `-1`.

}
```

After all variables are expanded, `widthC`'s value will be `calc(calc(100px / 2) / 2)`. When it's assigned to `.foo`'s width property, all inner `calc()` functions (no matter how deeply nested) will be flattened to just parentheses. Therefore, the `width` property's value will eventually be `calc((100px / 2) / 2)`, which equals `25px`. In short, a `calc()` inside of a `calc()` is identical to using parentheses.

> ### Adjusting colors dynamically
> 
> The `calc()` function can be used with colors to dynamically adjust individual color channels. In the example below, CSS variables are defined for the `red`, `green`, and `blue` color channels. The `calc()` function is used to dynamically decrease the `red` channel by `50` units.
> 
> ```html hidden
> <p class="original">This text color is the original color.</p>
> <p class="plus">
>   This text color is adjusted to increase the blue channel by 100.
> </p>
> <p class="minus">
>   This text color is adjusted to decrease the red channel by 200.
> </p>
> <p class="double">This text color is adjusted to double the green channel.</p>
> <p class="half">
>   This text color is dynamically adjusted to halve the alpha channel.
> </p>
> ```
> 
> ```css
> :root {
>   --r: 255;
>   --g: 100;
>   --b: 100;
>   --a: 1;
> }
> 
> .original {
>   color: rgb(var(--r) var(--g) var(--b));
> }
> 
> .plus {
>   color: rgb(var(--r) var(--g) calc(var(--b) + 100));
> }
> 
> .minus {
>   color: rgb(calc(var(--r) - 200) var(--g) var(--b));
> }
> 
> .double {
>   color: rgb(var(--r) calc(var(--g) * 2) var(--b));
> }
> 
> .half {
>   color: rgb(var(--r) var(--g) var(--b) / calc(var(--a) / 2));
> }
> ```
> 
> {{EmbedLiveSample('Adjusting_colors_dynamically', '700', '200')}}
> 
> Refer to the [Using math functions](/en-US/docs/Web/CSS/CSS_colors/Relative_colors#using_math_functions) section to see how the `calc()` function is used with relative colors.

Reviewer 2: What do you think about this?

* un-hide the HTML
* Make the variables more explicit (`--red` instead of `--r`)
* Reduce repeated text in example
* Make the `<p>` text a little larger

````suggestion
### Adjusting colors with variables

The `calc()` function can be used with colors to adjust individual color channels.
In the example below, CSS variables are defined for the red, green, blue, and alpha channels of the [`rgb()`](/en-US/docs/Web/CSS/color_value/rgb) color function.
A base color is defined using these variables, and in each paragraph, `calc()` is used to expand or dynamically adjust the variables for each color channel.

```html
<p class="original">This text has the original color.</p>
<p class="plus">Increasing the blue channel by 100.</p>
<p class="minus">Decreasing the red channel by 200.</p>
<p class="double">Doubling the amount of green in the original color.</p>
<p class="half">Halving the alpha (transparency) channel.</p>
```

```css hidden
p {
  font-family: monospace;
  font-size: 16px;
}
```

```css
:root {
  --red: 255;
  --green: 100;
  --blue: 100;
  --alpha: 1;
}

.original {
  color: rgb(var(--red) var(--green) var(--blue));
}

.plus {
  color: rgb(var(--red) var(--green) calc(var(--blue) + 100));
}

.minus {
  color: rgb(calc(var(--red) - 200) var(--green) var(--blue));
}

.double {
  color: rgb(var(--red) calc(var(--green) * 2) var(--blue));
}

.half {
  color: rgb(var(--red) var(--green) var(--blue) / calc(var(--alpha) / 2));
}
```

{{EmbedLiveSample('adjusting_colors_with_variables', '700', '200')}}
````

Reviewer 2: I feel like we can also add a relative colors example seeing as we're adding the syntax here, and it's a good comparison to include. What do you think about this:

````md
### Adjusting color channels in relative colors

The `calc()` function can adjust individual color channels in relative colors without the need for storing color channel values as variables.
In the example below, the first paragraph uses a [named color](/en-US/docs/Web/CSS/named-color).
In the three paragraphs that follow, `calc()` is used with the [`rgb()`](/en-US/docs/Web/CSS/color_value/rgb) and [`hsl()`](/en-US/docs/Web/CSS/color_value/hsl) functions to adjust the values of each color component relative to the original named color.

```html
<p class="original">This text has the original color.</p>
<p class="plus-blue">Increasing the blue channel by 40.</p>
<p class="semi-transparent">Halving the alpha (transparency) channel.</p>
<p class="lower-saturation">Decreasing the saturation by 40.</p>
<p class="inverting-hue">Adding 180 degrees to the hue angle.</p>
```

```css hidden
p {
  font-family: monospace;
  font-size: 16px;
}
```

```css
.original {
  color: cornflowerblue;
}

.plus-blue {
  color: rgb(from cornflowerblue r g calc(b + 40));
}

.semi-transparent {
  color: rgb(from cornflowerblue r g b / calc(alpha / 2));
}

.lower-saturation {
  color: hsl(from cornflowerblue h calc(s - 40) l);
}

.inverting-hue {
  color: hsl(from cornflowerblue calc(h + 180) s l);
}
```

{{EmbedLiveSample('adjusting_color_channels_in_relative_colors', '700', '200')}}
````

Author: I'm going to delete the example I added. It is flawed because it did not really demo the color channels in `calc()`, which was what my intention. Sorry for the work you put in to improve it.

We definitely should add an example to show the use of `calc()` in relative colors, which is what you've suggested. I've taken your suggestion and modified it a bit (like used a different color and changed some values).

}
```

After all variables are expanded, `widthC`'s value will be `calc(calc(100px / 2) / 2)`. When it's assigned to `.foo`'s width property, all inner `calc()` functions (no matter how deeply nested) will be flattened to just parentheses. Therefore, the `width` property's value will eventually be `calc((100px / 2) / 2)`, which equals `25px`. In short, a `calc()` inside of a `calc()` is identical to using parentheses.

### Adjusting color channels in relative colors

The `calc()` function can be used to adjust individual color channels in [relative colors](/en-US/docs/Web/CSS/CSS_colors/Relative_colors) without the need for storing color channel values as variables.

In the example below, the first paragraph uses a [named color](/en-US/docs/Web/CSS/named-color).
In the paragraphs that follow, `calc()` is used with the [`rgb()`](/en-US/docs/Web/CSS/color_value/rgb) and [`hsl()`](/en-US/docs/Web/CSS/color_value/hsl) functions to adjust the values of each color channel relative to the original named color.

```html

> <p class="original">Original text color in <code>rebeccapurple</code></p>
> <p class="same-original">
>   Same as original color, but using <code>calc()</code>
> </p>
> <p class="increase-green">Green channel increased by 180</p>
> <p class="reduce-alpha">Alpha channel (transparency) reduced by half</p>
> <p class="reduce-saturation">Saturation reduced by 60</p>
> <p class="invert-hue">Hue angle increased by 180 degrees</p>

Author: ```suggestion
<p class="original">Original text color in rebeccapurple</p>
<p class="increase-hue">Hue increased by 80</p>
<p class="increase-lightness">Lightness increased by 20</p>
<p class="decrease-lightness">Lightness decreased by 10</p>
```

```

```css hidden
p {
  font-family: monospace;
  font-size: 16px;
}
```

```css

> .original {
>   color: rebeccapurple;
> }
> 
> .same-original {
>   color: rgb(from rebeccapurple calc(r) calc(g) calc(b));
> }
> 
> .increase-green {
>   color: rgb(from rebeccapurple r calc(g + 180) b);
> }
> 
> .reduce-alpha {
>   color: rgb(from rebeccapurple r g b / calc(alpha / 2));
> }
> 
> .reduce-saturation {
>   color: hsl(from rebeccapurple h calc(s - 60) l);
> }
> 
> .invert-hue {
>   color: lch(from aquamarine l c calc(h + 180));
> }

Author: ```suggestion
.original {
  color: rebeccapurple;
}

.increase-hue {
  color: lch(from rebeccapurple l c calc(h + 80));
}

.increase-lightness {
  color: lch(from rebeccapurple calc(l + 20) c h);
}

.decrease-lightness {
  color: lch(from rebeccapurple calc(l - 10) c h);
}
```

>   color: hsl(from cornflowerblue calc(h + 180) s l);

Reviewer: Same issue. Querying the output color value via the [HTMLElement.style](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/style) property or the [CSSStyleDeclaration.getPropertyValue()](https://developer.mozilla.org/en-US/docs/Web/API/CSSStyleDeclaration/getPropertyValue) method returns the output color as a [color(srgb ...)](https://developer.mozilla.org/en-US/docs/Web/CSS/color_value/color) value. This will be really confusing if people use the devtools and hit control on the color swatch. The relative color tutorial covers this and has some good examples to inspire.

>   color: hsl(from cornflowerblue calc(h + 180) s l);

Reviewer 2: ```suggestion
  color: lch(from aquamarine l c calc(h + 180));
```
